### DISCUSSION TITLE
ESP32-CAM build with i2c support

### DISCUSSION THREAD

**Discussion by KaiBuesing** [arendst/Tasmota#18705] [Q&A] [answered]
I'm just getting desperate here. Maybe someone of you can help. I am trying to compile a build for an ESP32 cam. What I need is the I2C support and drivers for the BME680 and BH1750. I have modified the "user_config_override.h" accordingly. This is also executed because the wifi data fit. But what is not to find in the settings is i2c SCL/SDA. 
Someone has a tip for me ?
<img width="744" alt="image" src="https://github.com/arendst/Tasmota/assets/113059963/77c92e85-262a-4185-83bc-d2074326478b">
<img width="616" alt="image" src="https://github.com/arendst/Tasmota/assets/113059963/bc9e346d-d7a6-49db-ac3c-d0f97efc4f36">

**Comment by barbudor** [comment]
![Screenshot_2023-05-21-15-06-02-011_com android chrome-edit](https://github.com/arendst/Tasmota/assets/2996042/9de9ae8e-6e55-4956-9769-f1ecc9565bee)

In da docs
https://tasmota.github.io/docs/Compile-your-build/#customize-your-build

**Reply by KaiBuesing** [reply]
okay, i understand. If i would take tasmota32, which parts must be activated to have full ESP32-cam support ?

**Reply by barbudor** [reply]
Definitions of FIRMWARE_WEBCAM are here : https://github.com/arendst/Tasmota/blob/a76ebaae48bc7ec573fa277c1c9bed6e810b2607/tasmota/include/tasmota_configurations_ESP32.h#L203C1-L208

But it's a little more complicated as you will also need to add a custom environment in `platformio_tasmota_cenv.ini` based on the tasmota32-webcam emvironment in order to modify the CPU frequency to 240MHz like here : https://github.com/arendst/Tasmota/blob/a76ebaae48bc7ec573fa277c1c9bed6e810b2607/platformio_tasmota_env32.ini#L60-L68

Still I don't understand your problem as you can see in my 1st link that I2C is supposed to be already included in tasmota32-webcam
I need to check that

**Reply by arendst** [reply, marked as answer]
Yes since a week (or so) I2C is included in webcam

**Reply by barbudor** [reply]
great
thanks Theo

**Reply by KaiBuesing** [reply]
Thanks for the quick reply Barbudor & Theo :-)

In the dvl branch , support of i2c is now included , great !

But next stupid question from me - how to get support of the BME680 & BH1750 sensors in an easy way ?

By the way, i love this project. Great and fantastic work from all involved !

**Reply by barbudor** [reply]
Unfortunately as BMP and BH1750 are explicily disabled from `tasmota32-webcam` (see the #undef in FIRMWARE_WEBCAM above), you are back to square one and have to follow my initial instructions:

Add an evironement in `platformio_tasmota_cenv` such as:
```
 [env:tasmota32-kaiwebcam] 
 extends                 = env:tasmota32_base 
 board                   = esp32-fix 
 board_build.f_cpu       = 240000000L 
 build_flags             = ${env:tasmota32_base.build_flags} 
                           -DCAMERA_MODEL_AI_THINKER 
                           -DOTA_URL='"http://ota.tasmota.com/tasmota32/release/tasmota32-kaiwebcam.bin"' 
 lib_extra_dirs          = lib/lib_ssl, lib/libesp32 
```
(see that I have removed `-DFIRMWARE_WEBCAM` and name the environment by your first name)

And add those in the `user_config_override.h`:
```
#define USE_WEBCAM
#define ENABLE_RTSPSERVER
#define USE_SPI
#define USE_SDCARD
#define USE_I2C
#define USE_BMP
#define USE_BH1750
```

Then build `tasmota32-kaiwebcam` and that should be ok

**Reply by barbudor** [reply]
Just for you to understand
The build workflow includes in that order:
```
my_user_config.h
user_config_override.h
tasmota_configurations.h or tasmota_configurations_ESP32.h
```

This is why defines done in `user_config_override.h` may be undef by variants FIRMWARE_XXX

**Reply by KaiBuesing** [reply]
Thanks for your effort and patience with me! I tried it according to your instructions, but I get an error message "fatal error: IRremoteESP8266.h: No such file or directory" when compiling. Any idea ? 
<img width="1092" alt="image" src="https://github.com/arendst/Tasmota/assets/113059963/b6211bf3-bcdd-44ae-8b5b-9f908dc88b91">
<img width="925" alt="image" src="https://github.com/arendst/Tasmota/assets/113059963/a295a694-6087-4038-8b0c-1ac270099719">

**Reply by barbudor** [reply]
Sorry for late reply

I can reproduce
I'm investigating

**Reply by barbudor** [reply]
Ok, it was not difficult in fact
You need to copy in your `user_config_override.h` all of the FIRMWARE_WEBCAM 
and replace the #undef USE_BMP and #undef USE_BH1750 by #define USE_BMP and #define USE_BH1750

the default `my_user_config.h` includes a lot of things that you don't need

So `user_config_overide.h` should looks like:
```
#define USE_WEBCAM
#define ENABLE_RTSPSERVER
#define USE_SPI
#define USE_SDCARD

#define USE_I2C
#undef USE_HOME_ASSISTANT
#define USE_TASMOTA_DISCOVERY                    // Enable Tasmota Discovery support (+2k code)
#undef USE_DOMOTICZ
#undef USE_SERIAL_BRIDGE
#undef ROTARY_V1
#undef USE_IR_REMOTE
#undef USE_ADC
#undef USE_AC_ZERO_CROSS_DIMMER
#undef USE_PWM_DIMMER
#undef USE_PWM_DIMMER_REMOTE
#undef USE_TUYA_MCU
#undef USE_EMULATION_HUE
#undef USE_EMULATION_WEMO
#undef USE_BUZZER
#undef USE_ARILUX_RF
#undef USE_DS18x20
#define USE_BMP                                    // <<== include BMP family
#undef USE_DHT
#define USE_BH1750                                // <<== include BH1750
#undef USE_WS2812
#undef USE_ENERGY_SENSOR
#undef USE_SHUTTER
#undef USE_DEVICE_GROUPS
//#undef USE_BERRY                                 // Disable Berry scripting language
#undef USE_MI_ESP32                             // (ESP32 only) Disable support for ESP32 as a BLE-bridge (+9k2 mem, +292k flash)
#undef USE_BLE_ESP32
```

**Reply by KaiBuesing** [reply]
Hi Barbudor,
no reason to apologize. I think most of us still have a life where they have to earn money and then can not take care of your hobby ;-)

With the "user_config_overide.h" you suggested, compiling works, but the BME680 is not supported. But an i2cscan shows that it is found on the bus. I guess this is because of the missing "#define USE_BME68X".

When I try to compile this with it, I get an error again. Do you have another idea.

Thanks a lot for your effort !  
<img width="991" alt="image" src="https://github.com/arendst/Tasmota/assets/113059963/ade08f5e-9326-4a24-80c0-f830fd3c98a4">
<img width="800" alt="image" src="https://github.com/arendst/Tasmota/assets/113059963/f7b77312-4802-477b-86e8-4e54b32b7012">

**Reply by Jason2866** [reply]
@KaiBuesing Probably missing of including the needed libs. Try this
```
 [env:tasmota32-kaiwebcam] 
 extends                 = env:tasmota32_base 
 board                   = esp32-fix 
 board_build.f_cpu       = 240000000L 
 build_flags             = ${env:tasmota32_base.build_flags} 
                           -DCAMERA_MODEL_AI_THINKER 
                           -DOTA_URL='""' 
 lib_extra_dirs          = lib/libesp32, lib/libesp32_lvgl, lib/lib_basic, lib/lib_i2c, lib/lib_rf, lib/lib_div, lib/lib_ssl, lib/lib_display, lib/lib_audio
```

**Reply by KaiBuesing** [reply]
Hi Jason,

thanks for your quick reply an hint !

But now i get the following error message.
<img width="1460" alt="image" src="https://github.com/arendst/Tasmota/assets/113059963/aae64f20-d001-43b4-b71b-ed8cc1f44ed8">

**Reply by Jason2866** [reply]
@KaiBuesing You have an whitespace before `lib_extra_dirs` this is doing harm.

**Reply by KaiBuesing** [reply]
You are right. Verry stupid mistake 😉

Now it works.

Many thanks for your support !!

<img width="526" alt="image" src="https://github.com/arendst/Tasmota/assets/113059963/e7ca353e-cbb2-4e70-8bae-b71a2789400f">

**Reply by mischa-t** [reply]
Hello Kai,
i was also intereseted in the esp cam with temperature functions.
Until now i have not managed to get it done.
Would you share you final FW?
Greetings Mischa

**Reply by thomasabbott** [reply]
@Jason2866 From another struggling esp user - Thank You for this thread!
I wanted to add USE_DS18x20 to my tasmota32-webcam build, and commenting out the #undef left all sorts of dependecy issues.
With your kaiwebcam method, it builds, it works!
![image](https://github.com/user-attachments/assets/f37d403d-4a8f-4fc5-91ee-594ab38e5736)
